feat: create endpoint to return projects by activity [WEB-676]

julian-determined-ai opened the pull request:
## Description

feat: create endpoint to return projects by activity [WEB-676]

Adds an endpoint to return all projects, sorted by `activity_time` that a user has accessed based on the `activity` table.
<!---
Lead with the intended commit body in this description field. For breaking
changes, please include "BREAKING CHANGE:" at the beginning of your commit
body.  At a minimum, this section should include a bracketed reference to the
Jira ticket, e.g. "[DET-1234]". When squash-and-merging, copy this directly
into the description field.
-->



## Test Plan

1. Use the swagger API to create an activity entry for a project and user.
2. Use the swagger API to make a get request to the new endpoint, only the project with an entry from (1) should be returned.
3. Use the swagger API to create another activity entry for a project and user.
4. Use the swagger API to make a get request to the new endpoint, the two projects should be returned now. The project from (3) should appear first in the list.
5. Use the swagger API to create another entry for the same project as (1)
6. Use the swagger API to make a get request to the new endpoint, the two projects should be returned now. The project from (5) should appear first in the list.

<!---
Describe the situations in which you've tested your change, and/or a screenshot
as appropriate.  Reviewers may ask questions about this test plan to ensure
adequate manual coverage of changes.
-->



## Commentary (optional)

<!---
Use this section of your description to add context to the PR. Could be for
particularly tricky bits of code that could use extra scrutiny, historical
context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->



## Checklist

- [x] Changes have been manually QA'd
~- [ ] User-facing API changes need the "User-facing API Change" label.~
~- [ ] Release notes should be added as a separate file under `docs/release-notes/`.~
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
~- [ ] Licenses should be included for new code which was copied and/or modified from any external code.~

<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:

- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:

- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:

- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")

-->


[WEB-676]: https://determinedai.atlassian.net/browse/WEB-676?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ioga commented on master/internal/api_project.go:
```diff
@@ -345,3 +345,57 @@ func (a *apiServer) UnarchiveProject(
+	  LEFT JOIN workspaces AS w on w.id = p.workspace_id
+	GROUP BY p.user_id, p.id, p.name, p.workspace_id, p.description, 
+	p.immutable, p.notes, p.state, p.error_message,  u.username, w.name, p.activity_time
+	ORDER BY p.activity_time DESC;`).
```
how do we control the size of the `activity` table? say I am using the product for an extended period of time, touching hundreds of projects over a couple years. 
- should we cleanup old activity?
- should we limit the number of returned projects here to some number, or have pagination for this API?

julian-determined-ai replied:
cc: @gt2345 since she will be taking over from here. 

I do think we should support pagination in some form we will only ever be showing a fixed number (I am not sure if we will be allowing users to change this number manually). I think we might only even need a limit and not necessarily an offset since we might just be showing a fixed size list.

I'd have to recheck with @alexjdetermined but I think we always want a user to see their "most recent" activity whether that was 1 minute or 1 year ago. So we might not be able to delete "old" entries. But instead we'd need some logic such as "keep at least X number of the most recent entries" so the dashboard can always be populated fully. 

I do think pagination should be implemented in some form in the PR even if it is a fixed size limit.

ioga replied:
I see. if there's no offset support then probably real pagination is not necessary.

julian-determined-ai commented:
@ioga @gt2345 @johnkim-det 

I have added a limit that is optional to the query. For now we do not believe we need to support an offset. This implementation expects the front end to send a limit with each request, and I have implemented a check to make sure we do not exceed the `MaxLimit` we defined as "500".
